### PR TITLE
Add mini-guide on implementing primitives

### PR DESCRIPTION
**Pull request by scolsen**
Just a small walkthrough on the common steps for implementing a new primitive.

### CODE REVIEW

File: docs/hacking.md [hunk 1]
```diff
@@ -176,3 +176,150 @@ files that serve different purposes in the compiler:
...
+high-level steps. The following sections provide some guidance on making such
+changes.
+
+### Adding a new Primitive
+
+If it doesn't require anything fancy or out of the ordinary, adding a new
+primitive to the compiler entails the following:
+
+1. Define your new primitive in `Primitives.hs`
+2. Add your primitive to the staring environment using `makePrim` in
```
Comment (hellerve) [inline docs/hacking.md:191]: Typo: `staring`->`starting`.

File: docs/hacking.md [hunk 2]
```diff
@@ -176,3 +176,150 @@ files that serve different purposes in the compiler:
...
+
+Primitives are functions of the `Primitive` type:
+
+```
+type Primitive = XObj -> Context -> [XObj] -> IO (Context, Either EvalError XObj)
+```
+
+Every primitive takes an xobj, the form that represents the primitive, a
+compiler context, and a list of XObjs the primitive form's arguments.
+Primitives return either a new `Context` updated based on the logic they
```
Comment (hellerve) [inline docs/hacking.md:204]: Clarification: Primitives always return a context, and either an `XObj` or an `EvalError`.

File: docs/hacking.md [hunk 3]
```diff
@@ -176,3 +176,150 @@ files that serve different purposes in the compiler:
...
+
+For example, here's how the `defmodule` primitive maps to the `Primitive` type:
+
+```
+(defmodule Foo (defn bar [] 1))
+ |         |-----------------|
+ XObj      [XObj] (arguments)
+```
+
+> The `Context` argument is a compiler concept and doesn't have a corresponding
```
Comment (hellerve) [inline docs/hacking.md:215]: Maybe add that it stands for the state of the compiler?

File: docs/hacking.md [hunk 4]
```diff
@@ -176,3 +176,150 @@ files that serve different purposes in the compiler:
...
+            return $ evalError ctx ("`immutable` expects a variable as an argument") (info xobj)
+  primitiveImmutable _ _ xobjs = -- any other number or types of xobj arguments are incorrect! Let's error.
+    return $ evalError ctx ("`immutable` expected a single symbol argument, but got" ++ show xobjs) (info xobj)
+  ```
+
+And that wraps up the core logic of our primitive. To make it available, we
+just need to register it in `StartingEnv.hs`.
+
+
+#### 1. Add your primitive to the starting environment
```
Comment (hellerve) [inline docs/hacking.md:314]: Why is this heading numbered?There do not seem to be any other numbers anywhere?

Comment (scolsen) [reply docs/hacking.md:314]: There is another numbered heading above, but I think it makes sense just to drop the numbers entirely (I numbered this one wrong anyways).

File: docs/hacking.md (hunk 3 above)
Comment (scolsen) [reply docs/hacking.md:215]: Thanks, done!

File: docs/hacking.md (hunk 2 above)
Comment (scolsen) [reply docs/hacking.md:204]: Ah good catch! Thank you!

File: docs/hacking.md (hunk 1 above)
Comment (scolsen) [reply docs/hacking.md:191]: Thanks!

**Comment by scolsen** [comment]
Thanks for taking a look, @hellerve 🙇

**Comment by hellerve** [comment]
Looks really great now, thank you!

**Comment by eriksvedang** [comment]
Such a nice read @scolsen . Very pedagogical!

**Pull request merged**